Voice lifecycle

jpcima opened the pull request:
Follow up on the numbering PR #245
It's a suggestion to be able to track the lifecycle of voices globally.

The purpose is to allocate and free voice-specific resources in relation with the mod matrix.

jpcima committed:
Numbering of regions and voices

jpcima committed:
Add a listener of voice state changes

paulfd commented on src/sfizz/Region.h:
```diff
@@ -237,6 +245,8 @@ struct Region {
      */
     float getGainToEffectBus(unsigned number) const noexcept;
 
+    const int regionNumber {};
```
Since the region still has a default constructor, should we initialize to -1, put an optional, or just disable the default constructor entirely?

paulfd commented on src/sfizz/Synth.cpp:
```diff
@@ -95,7 +102,8 @@ void sfz::Synth::onParseWarning(const SourceRange& range, const std::string& mes
 void sfz::Synth::buildRegion(const std::vector<Opcode>& regionOpcodes)
 {
-    auto lastRegion = absl::make_unique<Region>(resources.midiState, defaultPath);
+    int regionNumber = static_cast<int>(regions.size());
```
Regarding the above comment, the region number could also be an unsigned then

paulfd commented on src/sfizz/Voice.h:
```diff
@@ -255,12 +279,16 @@ class Voice {
+     */
+    void switchState(State s);
+
+    const int voiceNumber {};
```
Same comments

paulfd commented on src/sfizz/Voice.h:
```diff
@@ -31,14 +31,38 @@ class Voice {
+
+    class StateListener {
+    public:
+        virtual void onVoiceStateChanged(int /*idNumber*/, State /*state*/) {}
```
On the same idea, the region stores an active/inactive state. Would it be useful to have listeners for this in your applications or a further one?

paulfd replied:
Also thinking some more about it: wouldn't a pointer to `this` be more useful than the id number? or maybe both?

jpcima replied:
The id can be deduced from the Region/Voice, it's true, but on the other hand voice id number creates less coupling, and I don't think I necessarily require my future work to look into details of Voice or Region. (but that remains a thing to be confirmed, I'm not fully sure)

paulfd commented:
In the current process and if I remember correctly, the set of regions is not "immutable" per-say, since we build all of them and *then* we go through all the sample to i) check that they exist, and ii) preload them. This will disrupt the region index thing. If the default constructor of the regions is disabled I guess we could maybe check that the sample exists on creating the region? Or maybe disable the region rather than removing it from the set?

paulfd commented:
Just a small "up". I would use something like this for the polyphony groups too probably.

jpcima commented:
On this indexing remark, the idea is to retract this invariant that id=index.

Instead of direct indexing I consider to make it linear or binary search, where the search is initiated at index `id`.
In the very vast majority of times the id and index are identical and it's a direct hit, otherwise the id will be located somewhere in the near neighborhood.

paulfd commented:
I see! Ok then we can try it like this.

jpcima committed:
Initialize id number to -1, do not consider it as index

jpcima committed:
Implement the identifier in a better and more safe way

jpcima committed:
Update benchmarks

jpcima committed:
Add the lookup functions by identifier

jpcima committed:
Make region removal order-preserving

jpcima committed:
Add test

jpcima commented:
Rebased, tested, implemented lookup by ID.
This lead into the problem that removed Regions are out of order, due to the iter_swap way of doing.
Substituting with O(n) removal, the order is kept, and search works correctly in the presence of gaps.
